### PR TITLE
[DebugInfo][SIL] Introduce the 'implicit' attribute for debug variable

### PR DESCRIPTION
**Pull request by mshockwave**
Debug variables that are marked 'implicit' on its `debug_value` instruction mean that they were generated by compiler. Optimizers are free to remove them (if it becomes a dead code, for instance) even in  -Onone. Since they are barely used by users and keeping them might lead to incorrect IRGen results.

**NOTE**: This PR is only for code review. It depends on and _includes_  #38734 . **DO NOT MERGE**. Please merge #38736 instead when everything is done.

### CODE REVIEW

**Comment by mshockwave** [comment]
@swift-ci please test

File: include/swift/SIL/SILInstruction.h [hunk 1]
```diff
@@ -1796,16 +1805,60 @@ class TailAllocatedDebugVariable {
...
                               DeclLoc, DeclScope, DIExprElements);
+    else
+      return SILDebugVariable(getName(buf), isLet(), getArgNo(), isImplicit(),
+                              AuxVarType, DeclLoc, DeclScope, DIExprElements);
   }
 };
 static_assert(sizeof(TailAllocatedDebugVariable) == 4,
               "SILNode inline bitfield needs updating");
 
+class SILDebugVariableSupplement {
```
Comment (adrian-prantl) [inline include/swift/SIL/SILInstruction.h:1818]: See comments in other PR.

**Review by adrian-prantl — APPROVED** [review]

**Comment by mshockwave** [comment]
@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]
Looks good now. Can you add a test (in a separate PR) that tests what happens when we exceed name_length?

**Pull request merged**